### PR TITLE
QNTM-1635: JSON deserialization should create dummy nodes when type is not found

### PR DESCRIPTION
**Pull request by ColinDayOrg**
### Purpose

This PR implements the serialization/deserialization of dummy nodes in Dynamo when the actual node is not found.

### Declarations

Check these if you believe they are true

- [X] The code base is in a better state after this PR
- [n/a] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [X] The level of testing this PR includes is appropriate
- [n/a] User facing strings, if any, are extracted into `*.resx` files
- [X] All tests pass using the self-service CI.
- [n/a] Snapshot of UI changes, if any.
- [n/a] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions), and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.

### Reviewers

@mjkkirschner

### CODE REVIEW

File: src/DynamoCore/Graph/Workspaces/SerializationConverters.cs [hunk 1]
```diff
@@ -442,6 +458,52 @@ public override object ReadJson(JsonReader reader, Type objectType, object exist
         }
     }
 
+    /// <summary>
+    /// WorkspaceWriteConverter is used for serializing Workspaces to JSON.
```
Comment (mjkkirschner) [inline src/DynamoCore/Graph/Workspaces/SerializationConverters.cs]: comment is incorrect.

File: src/DynamoCore/Graph/Workspaces/SerializationConverters.cs [hunk 2]
```diff
@@ -80,21 +80,37 @@ public override object ReadJson(JsonReader reader, Type objectType, object exist
                 var mangledName = obj["FunctionSignature"].Value<string>();
 
                 var description = libraryServices.GetFunctionDescriptor(mangledName);
-
-                if (type == typeof(DSVarArgFunction))
+                if (description == null)
```
Comment (mjkkirschner) [inline src/DynamoCore/Graph/Workspaces/SerializationConverters.cs]: I'm not understanding this - why do we only create dummy nodes in the case that the description is null? - Or am I missing reading the code - I would assume we do it anytime the Type above is null on line 54.

File: src/DynamoCore/Graph/Workspaces/SerializationConverters.cs [hunk 3]
```diff
@@ -442,6 +458,52 @@ public override object ReadJson(JsonReader reader, Type objectType, object exist
...
+    /// WorkspaceWriteConverter is used for serializing Workspaces to JSON.
+    /// </summary>
+    public class DummyNodeWriteConverter : JsonConverter
+    {
+        //private EngineController engine;
+
+        public DummyNodeWriteConverter(EngineController engine = null)
+        {
+            //if (engine != null)
+            //{
```
Comment (mjkkirschner) [inline src/DynamoCore/Graph/Workspaces/SerializationConverters.cs]: whats this?

File: src/DynamoCore/Graph/Workspaces/SerializationExtensions.cs [hunk 4]
```diff
@@ -29,7 +29,8 @@ internal static string ToJson(this WorkspaceModel workspace, EngineController en
                 Formatting = Formatting.Indented,
                 Converters = new List<JsonConverter>{
                         new ConnectorConverter(),                        
-                        new WorkspaceWriteConverter(engine)
+                        new WorkspaceWriteConverter(engine),
+                        new DummyNodeWriteConverter()
```
Comment (mjkkirschner) [inline src/DynamoCore/Graph/Workspaces/SerializationExtensions.cs]: what happens with view serialization, does the nodeView of a dummy node also get serialized correctly? I think it would be good to have minimal automated testing thats interact with dummy nodes. That will take some messing about because the serialization tests ignore graphs that load up with dummy nodes.

File: src/DynamoCore/Graph/Workspaces/SerializationConverters.cs [hunk 5]
```diff
@@ -442,6 +458,52 @@ public override object ReadJson(JsonReader reader, Type objectType, object exist
...
+    /// <summary>
+    /// WorkspaceWriteConverter is used for serializing Workspaces to JSON.
+    /// </summary>
+    public class DummyNodeWriteConverter : JsonConverter
+    {
+        //private EngineController engine;
+
+        public DummyNodeWriteConverter(EngineController engine = null)
+        {
+            //if (engine != null)
```
Comment (ramramps) [inline src/DynamoCore/Graph/Workspaces/SerializationConverters.cs]: engine is required here?

**Comment by mjkkirschner** [comment]
@ColinDayOrg there is a merge conflict in this PR.
otherwise looks like a good start but I don't understand why the dummy constructor is invoked where it is. - I also think this needs some automated testing.

File: src/DynamoCore/Graph/Workspaces/SerializationExtensions.cs [hunk 6]
```diff
@@ -29,7 +29,8 @@ internal static string ToJson(this WorkspaceModel workspace, EngineController en
                 Formatting = Formatting.Indented,
                 Converters = new List<JsonConverter>{
                         new ConnectorConverter(),                        
-                        new WorkspaceWriteConverter(engine)
+                        new WorkspaceWriteConverter(engine),
```
Comment (ramramps) [inline src/DynamoCore/Graph/Workspaces/SerializationExtensions.cs]: different question here :  any reason for not using WorkspaceWriteConverter? WriteJSON method can validate this and take a different path. I could be totally wrong here.

File: src/DynamoCore/Graph/Workspaces/SerializationConverters.cs (hunk 2 above)
Comment (ColinDayOrg) [reply src/DynamoCore/Graph/Workspaces/SerializationConverters.cs]: My testing is using a zero-touch node package. After uninstalling the package and loading my test graph I see that the type value is not null, even though the library is not available. Getting the description invokes the library services, which returns null because the library is not loaded. There may be a more direct way of making the query, but this call was fairly similar to the code that is called when loading XML

File: src/DynamoCore/Graph/Workspaces/SerializationConverters.cs (hunk 1 above)
Comment (ColinDayOrg) [reply src/DynamoCore/Graph/Workspaces/SerializationConverters.cs]: Copy/paste error. The comment has been updated to be correct.

File: src/DynamoCore/Graph/Workspaces/SerializationConverters.cs (hunk 2 above)
Comment (mjkkirschner) [reply src/DynamoCore/Graph/Workspaces/SerializationConverters.cs]: If the type is not actually loaded into the app domain, it should be null when getType searches for it, it's possible you still have the type in memory but the library is not loaded.

Comment (mjkkirschner) [reply src/DynamoCore/Graph/Workspaces/SerializationConverters.cs]: continued - my point is this is a case we need to handle where type is really null.

File: src/DynamoCore/Graph/Workspaces/SerializationConverters.cs (hunk 5 above)
Comment (ColinDayOrg) [reply src/DynamoCore/Graph/Workspaces/SerializationConverters.cs]: This was a vestigial remnant from copying another converter. Engine is not needed here, or the default constructor, I have removed both.

File: src/DynamoCore/Graph/Workspaces/SerializationExtensions.cs (hunk 6 above)
Comment (ColinDayOrg) [reply src/DynamoCore/Graph/Workspaces/SerializationExtensions.cs]: I went with a DummyNodeConverter because the WorkspaceConverter serializes the nodes as a collection, so it seemed to me to be a clean way to properly serialize a DummyNode to it's original content.

File: src/DynamoCore/Graph/Workspaces/SerializationExtensions.cs (hunk 4 above)
Comment (ColinDayOrg) [reply src/DynamoCore/Graph/Workspaces/SerializationExtensions.cs]: The view serialization appears to work as expected, but there needs to be some more testing done. I will look into this more closely.

File: src/DynamoCore/Graph/Workspaces/SerializationConverters.cs (hunk 2 above)
Comment (mjkkirschner) [reply src/DynamoCore/Graph/Workspaces/SerializationConverters.cs]: @ColinDayOrg I think I understand whats happening:
zerotouch nodes share a base class which is always going to be loaded so that is why type is not null for this node in your package.

We need to test a nodeModel type which is null so that will be a non zero touch node.

Also, I'm still not sure a null description will work for zero touch node - couldn't it just be a zero touch node with no description? Also for zero touch nodes we'll need to handle the base class and the dsvararg class (variable inputs)

Also have you considered customNodes? (Function) class?

File: src/DynamoCore/Graph/Workspaces/SerializationConverters.cs [hunk 7]
```diff
@@ -63,18 +63,30 @@ public override object ReadJson(JsonReader reader, Type objectType, object exist
...
+                node = function;
+
+                if (isUnresolved)
+                  function.UpdatePortsForUnresolved(inPorts, outPorts);
             }
             else if (type == typeof(CodeBlockNodeModel))
             {
                 var code = obj["Code"].Value<string>();
                 node = new CodeBlockNodeModel(code, guid, 0.0, 0.0, libraryServices, ElementResolver);
-                RemapPorts(node, inPorts, outPorts, resolver);
```
Comment (mjkkirschner) [inline src/DynamoCore/Graph/Workspaces/SerializationConverters.cs]: this is not required because of nodemodel changes?

Comment (ColinDayOrg) [reply src/DynamoCore/Graph/Workspaces/SerializationConverters.cs]: The only time that calling RemapPorts is not required is when there is a JSON constructor that forwards the ports. There were many calls to RemapPorts that have been consolidated to after the big if/elseif/.../else statement to simplify the code. (In other words the call to RemapPorts for CodeBlockNodeModel nodes is still being made, just in a different place. See line 134.)

File: src/DynamoCore/Core/CustomNodeManager.cs [hunk 8]
```diff
@@ -135,34 +135,97 @@ protected virtual void OnCustomNodeRemoved(Guid functionId)
...
+                return true;
             }
-            else
+
+            // Couldn't get the workspace with the given ID, try a name lookup instead.
+            if (name != null && !TryGetNodeInfo(name, out info))
+                return false;
+
+            // Try to get the definition using the function ID, initializing the custom node if necessary
+            if (info != null && TryGetFunctionDefinition(info.FunctionId, isTestMode, out def))
```
Comment (QilongTang) [inline src/DynamoCore/Core/CustomNodeManager.cs]: @ColinDayOrg It seems `TryGetFunctionDefinition()` and the code path underneath are still based on xml doc, do you see todos while developing? Just curious.

Comment (ColinDayOrg) [reply src/DynamoCore/Core/CustomNodeManager.cs]: It does appear that there are multiple issues with custom nodes around XML versus JSON (some of which I believe Sriram just addressed), but it seemed somewhat out of scope to me for the current task. I would expect that once the dummy node code is merged that there would be a fair amount of testing that will expose some issues that would be logged as bugs and addressed in the future. If you think it would be more appropriate I can add "TODO" comments.

Comment (QilongTang) [reply src/DynamoCore/Core/CustomNodeManager.cs]: That's fine. I was just curious anything you noticed as known issues.

File: src/DynamoCore/Graph/Workspaces/SerializationConverters.cs (hunk 2 above)
Comment (ColinDayOrg) [reply src/DynamoCore/Graph/Workspaces/SerializationConverters.cs]: After attempting to redo this check with something more specific (i.e. mirroring the code from the ZeroTouchNodeLoader) it became apparent that there needs to be some more infrastructure work to be able to utilize the new code. 

The two main reasons are;

1. In the XML format there is a field for zero-touch nodes that contains the assembly that the zero-touch node is in. This attribute is not currently present in the JSON format and will need to be added. This missing information is a blocking issue for updating this check to use the more specific code.
2. The name of the node has been moved to the view properties and is not available at the time the node model is loaded. This may not be an issue as the name appears to be used to identify built-in node types, and it is not clear if this is important at the point where we are loading zero-touch nodes.

I think that we may be able to use the more specific code once the assembly information is saved in the JSON format.

Comment (mjkkirschner) [reply src/DynamoCore/Graph/Workspaces/SerializationConverters.cs]: @ColinDayOrg the assembly is saved as part of the concrete type like this `typname, assembly`

Comment (ColinDayOrg) [reply src/DynamoCore/Graph/Workspaces/SerializationConverters.cs]: I should have been more specific in my previous comment, I was talking about the path to the assembly, not the name of the assembly. 

In the case of the XML file format the assembly is saved as the path to the assembly; "C:\Users\dayc\AppData\Roaming\Dynamo\Dynamo Core\1.2\packages\DynamoPrimitives\bin\DynamoPrimitives.dll")

In the case of the JSON file format it is saved in a type name format; "Dynamo.Graph.Nodes.ZeroTouch.DSFunction").

Comment (mjkkirschner) [reply src/DynamoCore/Graph/Workspaces/SerializationConverters.cs]: so we discussed this hintpath as part of another task, what code are you talking about it being required for? I can't remember the outcome of that discussion, I think we decided we could add hintpath as a view layer property since it will have no meaning outside of dynamo.

Comment (ColinDayOrg) [reply src/DynamoCore/Graph/Workspaces/SerializationConverters.cs]: The assembly information is used in ZeroTouchNodeLoader.cs, line 76-100, to retrieve the FunctionDescriptor for the given function. This appears to me to be the best way to determine if the zero-touch node can be resolved. Currently it seems that there is not enough information saved in the JSON format to do a formal retrieval of the needed information.

Comment (mjkkirschner) [reply src/DynamoCore/Graph/Workspaces/SerializationConverters.cs]: I think we'll need to rethink this based on the other task. presently you make a guess based on the function descriptor being null?

Comment (mjkkirschner) [reply src/DynamoCore/Graph/Workspaces/SerializationConverters.cs]: I know it's a bit of a style thing, but can you rename description to functionDescriptor or something like that so it's not confused with node description text.

Comment (ColinDayOrg) [reply src/DynamoCore/Graph/Workspaces/SerializationConverters.cs]: I have renamed the variable, I agree it is clearer. 

The choice to use the function descriptor being null as a switch was based on the code in ZeroTouchNodeLoader.cs where the same test is used to determine (when loading XML) if a dummy node should be created or not. Note though that the test in the zero-touch loader is slightly more complex, but uses information that is not currently available from the JSON format.
